src/ppx/scene/scene_node.cpp

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ppx/scene/scene_node.h"
+
+namespace ppx {
+namespace scene {
+
+// -------------------------------------------------------------------------------------------------
+// Node
+// -------------------------------------------------------------------------------------------------
+Node::Node(scene::Scene* pScene)
+    : mScene(pScene)
+{
+}
+
+Node::~Node()
+{
+}
+
+void Node::SetVisible(bool visible)
+{
+    mVisible = visible;
+}
+
+const float4x4& Node::GetEvaluatedMatrix() const
+{
+    if (mEvaluatedDirty) {
+        float4x4 parentEvaluatedMatrix = float4x4(1);
+        if (!IsNull(mParent)) {
+            parentEvaluatedMatrix = mParent->GetEvaluatedMatrix();
+        }
+        auto& concatenatedMatrix = GetConcatenatedMatrix();
+        mEvaluatedMatrix         = parentEvaluatedMatrix * concatenatedMatrix;
+        mEvaluatedDirty          = false;
+    }
+    return mEvaluatedMatrix;
+}
+
+void Node::SetParent(scene::Node* pNewParent)
+{
+    mParent = pNewParent;
+}
+
+void Node::SetEvaluatedDirty()
+{
+    mEvaluatedDirty = true;
+    for (auto& pChild : mChildren) {
+        pChild->SetEvaluatedDirty();
+    }
+}
+
+void Node::SetTranslation(const float3& translation)
+{
+    Transform::SetTranslation(translation);
+    SetEvaluatedDirty();
+}
+
+void Node::SetRotation(const float3& rotation)
+{
+    Transform::SetRotation(rotation);
+    SetEvaluatedDirty();
+}
+
+void Node::SetScale(const float3& scale)
+{
+    Transform::SetScale(scale);
+    SetEvaluatedDirty();
+}
+
+void Node::SetRotationOrder(Transform::RotationOrder value)
+{
+    Transform::SetRotationOrder(value);
+    SetEvaluatedDirty();
+}
+
+scene::Node* Node::GetChild(uint32_t index) const
+{
+    scene::Node* pChild = nullptr;
+    if (ppx::GetElement(index, mChildren, &pChild)) {
```

Comment:
Should this negate the condition? `if (ppx::GetElement(index, mChildren, &pChild)`
`GetElement` returns true if the element was found?

Reply:
Also do we even need to check? Or just return pChild after calling GetElement?

Reply:
We don't. Check removed.